[RUBY-2704] Fix scoping issues in `Payment` model and update tests

jjromeo opened the pull request:
- Extend `Payment` class from `ApplicationRecord` for ActiveRecord functionality.
- Update `refunds_and_reversals` and `refundable` scopes to include `.success` for filtering successful payments.
- Add comprehensive tests for all payment scopes, ensuring correct filtering and ordering of payments.
- Refactor test setup for payment types and status to improve clarity and maintainability.

jjromeo committed:
[RUBY-2704] Fix scoping issues in `Payment` model and update tests

- Extend `Payment` class from `ApplicationRecord` for ActiveRecord functionality.
- Update `refunds_and_reversals` and `refundable` scopes to include `.success` for filtering successful payments.
- Add comprehensive tests for all payment scopes, ensuring correct filtering and ordering of payments.
- Refactor test setup for payment types and status to improve clarity and maintainability.

jjromeo committed:
Merge branch 'main' into RUBY-2704-scope-fix

jjromeo committed:
[RUBY-2704] Update SonarCloud action in CI workflow

Changed the GitHub action for SonarCloud analysis from `sonarcloud-github-action` to `sonarqube-scan-action` in the CI workflow configuration to ensure proper functionality.

jjromeo committed:
Merge branch 'RUBY-2704-scope-fix' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2704-scope-fix

jjromeo committed:
[RUBY-2704] Add `:success` trait to payment factory and update specs

- Introduced a `:success` trait in the payment factory to streamline the creation of successful payment objects.
- Updated various payment spec tests to utilize the new `:success` trait, simplifying the setup and ensuring consistency across tests.
- Removed redundant `success?` method stubbing in specs as the new trait inherently sets the `payment_status` to "success".

jjromeo committed:
[RUBY-2704] Remove unused `reverseable` scope from `Payment` model in WasteExemptionsEngine.

jjromeo committed:
[RUBY-2704] Add `:nocov:` comments to exclude scope methods from test coverage in `Payment` model

This change marks the `not_cancelled` and `refunds_and_reversals` scopes with `:nocov:` to exclude them from test coverage analysis.

jjromeo committed:
re-run sonarcloud

jjromeo committed:
Merge branch 'main' into RUBY-2704-scope-fix

jjromeo committed:
run sonarqube

jjromeo committed:
Merge branch 'RUBY-2704-scope-fix' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2704-scope-fix

jjromeo committed:
[RUBY-2704] Fix path substitution in CI workflow for coverage report

Updated the `sed` command in the CI workflow to use the `GITHUB_WORKSPACE` environment variable for path substitution, ensuring consistent path handling in the coverage report.

jjromeo committed:
[RUBY-2704] Update SonarCloud GitHub Action in CI workflow

Changed the SonarCloud action from `sonarsource/sonarqube-scan-action@master` to `sonarsource/sonarcloud-github-action@master` in the `.github/workflows/ci.yml` file to utilize the correct GitHub Action for SonarCloud analysis.

jjromeo committed:
[RUBY-2704] Fix sed command in GitHub Actions CI workflow to correctly replace GITHUB_WORKSPACE in coverage.json

jjromeo committed:
[RUBY-2704] Fix scope in CI workflow and update SonarCloud action

- Corrected the path replacement in the sed command for coverage report.
- Updated SonarCloud action to `sonarqube-scan-action@v4.1.0` for improved analysis.

jjromeo committed:
[RUBY-2704] Enable verbose logging for SonarCloud analysis in CI workflow

Added `-Dsonar.verbose=true` to the SonarCloud analysis step in the GitHub Actions CI workflow to provide more detailed logging output.

jjromeo committed:
[RUBY-2704] Fix path replacement in CI workflow for coverage report

Updated the `sed` command in `.github/workflows/ci.yml` to use the `GITHUB_WORKSPACE` environment variable for path replacement in `coverage.json`, ensuring correct path handling in the CI environment.

jjromeo committed:
[RUBY-2704] Fix coverage path in CI workflow and remove SonarCloud verbosity

Updated the path replacement in the coverage report to use the correct GitHub workspace path. Removed the verbose argument from the SonarCloud analysis step for cleaner logs.

jjromeo committed:
[RUBY-2704] Add debug step to CI workflow to list files in `app/models/concerns/` directory

This change introduces a debug step in the CI workflow to output the current directory and list files in the `app/models/concerns/` directory to aid in troubleshooting.

jjromeo committed:
[RUBY-2704] Fix coverage debug and update SonarCloud configuration

- Renamed the debug step in the CI workflow to `Debug Coverage` and added commands to print `GITHUB_WORKSPACE` and search for a specific term in `coverage.json`.
- Modified `sonar.sources` in `sonar-project.properties` by removing leading './' for consistency.

jjromeo committed:
[RUBY-2704] Fix coverage path in CI workflow

Updated the `sed` command in the CI workflow to correct the coverage path replacement in `coverage.json`, ensuring it uses `/github/workspace` instead of the hardcoded path.

jjromeo committed:
Fix SonarCloud integration in CI workflow

- Update path replacement in `sed` command for coverage report.
- Change SonarCloud GitHub action to `sonarsource/sonarqube-scan-action@master` for analysis.

jjromeo committed:
Merge branch 'sonar-cloud-fix' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2704-scope-fix

jjromeo committed:
Merge branch 'main' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2704-scope-fix

jjromeo committed:
[RUBY-2704] Fix sonar.sources paths in sonar-project.properties

Updated the `sonar.sources` property to use relative paths with `./` for consistency and accuracy in source file detection.

jjromeo committed:
[RUBY-2704] Remove unnecessary `:nocov:` comments from `Payment` model in `WasteExemptionsEngine` module.

PaulDoyle-EA commented on spec/models/waste_exemptions_engine/payment_spec.rb:
```diff
@@ -1,28 +1,29 @@
 # frozen_string_literal: true
 
 require "rails_helper"
+require WasteExemptionsEngine::Engine.root.join("app", "models", "waste_exemptions_engine", "payment")
```
Why is this necessary?

jjromeo replied:
It's not, added while trying to debug the sonarcloud issue! Removed.

PaulDoyle-EA commented on spec/models/waste_exemptions_engine/payment_spec.rb:
```diff
@@ -1,28 +1,29 @@
       context "when payment type is not refundable" do
-        let(:reversal_payment) { create(:payment, payment_type: Payment::PAYMENT_TYPE_REVERSAL, account:) }
+        before do
+          allow(payment).to receive(:payment_type).and_return(Payment::PAYMENT_TYPE_REVERSAL)
```
Can this be done in the `let` statement for the payment? Or as a trait in the factory?

jjromeo committed:
[RUBY-2704] Add `reverseable` scope to Payment model and refactor specs

- Introduced a new scope `reverseable` in the `Payment` model to identify payments that can be reversed, excluding certain types and associated payments.
- Refactored `maximum_refund_amount` tests in `Payment` spec to use `let` for `payment_type`, improving clarity and reducing redundancy.

jjromeo committed:
Merge branch 'main' into RUBY-2704-scope-fix
